L10-c++-reactive-streams.md

```diff
@@ -0,0 +1,387 @@
+
+
+## Rationale
+
+Alternate approaches for a C++ for gRPC include
+
+* **The current synchronous API:** Too slow for many use cases.
+* **The current asynchronous API:** Too low level for many use cases.
+* **Futures, as suggested in [issue #7352](https://github.com/grpc/grpc/issues/7352):** Futures typically do not have a concept of cancellation and they do not map naturally to streams.
+* **Callbacks, as suggested in [issue #7352](https://github.com/grpc/grpc/issues/7352):** Callbacks are rather low level; [Many node.js programmers would agree](http://callbackhell.com/) that it is not the nicest style of API. It is also tricky to avoid a `Send` method that never blocks and never eats all available memory.
+* **Synchronous API with fibers/lightweight threads:** This mitigates some of the performance issues of the synchronous API but it does not naturally allow the user of the API to do several things at once.
+
+The API proposed in this document has the following positive traits:
+
+* Encourages simple application code, even for applications that are not easy.
+* Fully asynchronous; efficient.
+* Supports not just unary RPCs but also streaming.
+* Supports transitive cancellation with little to no application code support.
+* Respects backpressure: When working with streams, there are no unbounded buffers, blocking or dreop or that can block or drop data.
+
+This API especially shines in applications that combine many async operations in complex ways:
+
+* Techniques such as optimistic retries for long tail latency mitigation are easy to express with Reactive Streams.
+* Application code that needs to do many upstream calls, some in parallel and others in sequence, some being streamed and some not, are helped greatly by Reactive Streams, because it allows the developer to not write the code otherwise required to manage the state that deals with orchestrating all of these calls.
+* Reactive Streams offers a clear and composable approach to writing asynchronous functions within a server as well: An asynchronous function is one that returns a Publisher. This can greatly help API consistency within a given programming language, not just on the wire protocol level, as well as making it easy to on a type level distinguish asynchronous code from synchronous code.
+* A concrete example of how code can benefit from Reactive Streams is [Netflix's Hystrix library](https://github.com/Netflix/Hystrix/wiki/How-it-Works), which helps provide latency and fault tolerance for microservices.
+
+The main drawbacks of this proposal are:
+
+* It is easy to use only after you have spent some time learning Reactive Streams' concepts: You must change mindset to thinking about data flows.
+* There are features in gRPC that don't obviously fit Reactive Streams, for example metadata.
```

Comment:
There needs to be a story for metadata.

Reply:
Ok. I will read and think about this more. This might take some time for me to fix however; this is more work than of your other comments so far.

Reply:
It might be worth looking at how grpc-java, grpc-objective-c deal with this, as they each have reactive stream allowances in their API's.

@ejona86, @jcanizales care to provide pointers?

Reply:
I will have a look. I don't think this will necessarily be a hard problem to solve. I have simply assumed that it should be possible somehow and not thought about it.

Reply:
grpc-java doesn't expose Metadata on the stubs; Metadata is only exposed to interceptors.

Reply:
@ctiller I have read a little bit about metadata and how interceptors work in grpc-java. I can see a few different directions that I could go designing this, and would appreciate advice on which I should try first. We could do more than one or even all of them at once, but for the sake of simplicity and MVP I would prefer trying only one at first if possible. The main directions I can think of are

1. Expose metadata directly to individual rpc handlers (aka what grpc-java does **not** do): This is probably the most conceptually lightweight approach and is appropriate if metadata is often used for RPC-specific purposes.
2. Introduce some kind of Reactive Streams style Interceptor API. This is good for doing middleware-style things such as auth, logging, metrics, tracing etc.
3. Add an Interceptor API to the async C++ API (perhaps it can even be done strictly as a layer outside of it). I think this could be done without adding too much complexity to the current async API. It might lead to lower level code in middleware than option 2. However, middlewares wouldn't need to depend on if the application uses Reactive Streams or not. It might even be possible to make the same Interceptor interface work with the sync API.

I like option number 3 even though it seems like the most amount of work, because I think middleware writers would benefit from it. What do you think?

Reply:
On Objective-C, request and response metadata are exposed as properties of the "call" object. This call object acts as an Rx-style source of response messages and a sink of request messages.